### PR TITLE
Refactor the quick-fix for the bins retrieval

### PR DESCRIPTION
**Pull request by alto**
We had to [quickly fix](https://github.com/sharesight/vimaly/pull/20) the list of all bins found on Vimaly, as the "Trade Confirmations" bin was not included anymore in the top 500 list of bins. This PR now cleans up the quick-n-dirty solution and improves test coverage and readability.

### CODE REVIEW

File: lib/vimaly/client.rb [hunk 1]
```diff
@@ -84,14 +84,25 @@ def bin(name_or_regex)
...
+
+        bins = []
+        current_page = 0
+        while bins.size < max_number_of_bins
+          # those request parameters are not documented, maybe this gets added at some
+          # point to https://vimaly.com/public/rest-help.html#Bins
+          chunk_of_bins = get("/bins?max-results=#{bins_per_request}&page-token=#{current_page * bins_per_request}")
+          current_page += 1
+          bins += chunk_of_bins
+          break if chunk_of_bins.size < bins_per_request
```
Comment (kylorhall) [inline lib/vimaly/client.rb:102]: So if you get back `0..499` bins it stops?  Makes sense to me.

File: lib/vimaly/client.rb [hunk 2]
```diff
@@ -84,14 +84,25 @@ def bin(name_or_regex)
...
-        data = (0..1).inject([]) do |list, page|
-          list += get("/bins?max-results=#{per_page}&page-token=#{page * 500}")
+        # 500 is the vimaly limit
+        bins_per_request = options[:bins_per_request] || 500
+        # 5000 is just a random value here; hopefully we never reach this limit
+        max_number_of_bins = options[:max_number_of_bins] || 5000
+
+        bins = []
+        current_page = 0
+        while bins.size < max_number_of_bins
```
Comment (kylorhall) [inline lib/vimaly/client.rb:96]: This makes sense too to avoid a leak 👍

File: test/client_test.rb [hunk 3]
```diff
@@ -175,21 +175,56 @@ class ClientTest < Minitest::Test
...
+        body: [{name: 'Gamma', _id: 3}, {name: 'Delta', _id: 4}].to_json
+      )
+      # 3rd page of 2 bins, returning only 1 (indicating the last page)
+      stub_request(:get, %r{#{Vimaly::Client::VIMALY_ROOT_URL}/rest/2/company_id/bins\?max-results=2\&page-token=4}).to_return(
+        status: 200,
+        body: [{name: 'Epsilon', _id: 5}].to_json
+      )
+    end
+
+    should "succeed and load all available bins" do
```
Comment (kylorhall) [inline test/client_test.rb:196]: Sweet, this test coverage LGTM, I think you're testing just about every scenario…barring a "failure scenario" wherein I think we just don't care if Vimaly goes boom?

**Review by kylorhall — APPROVED** [review]
👍

File: test/client_test.rb (hunk 3 above)
Comment (alto) [reply test/client_test.rb:196]: Well, I think if _vimaly_ goes 💥 this code here fails too, and we got notified right away. So no tests for that.

**Pull request merged**